src/Orleans.Core/Diagnostics/ActivityPropagationGrainCallFilter.cs

```diff
@@ -0,0 +1,108 @@
+        {
+            // rpc attributes from https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/semantic_conventions/rpc.md
+            ActivityTagsCollection tags = null;
+            if (activitySource.HasListeners())
+            {
+                tags = new ActivityTagsCollection
+                {
+                    {"rpc.service", context.InterfaceMethod?.DeclaringType?.FullName},
+                    {"rpc.method", context.InterfaceMethod?.Name},
+                    {"net.peer.name", context.Grain?.ToString()},
+                    {"rpc.system", "orleans"}
+                };
+            }
+
+            using var activity = activitySource.StartActivity(activityName, activityKind, activityContext, tags);
+            if (activity is not null)
+            {
+                RequestContext.Set(TraceParentHeaderName, activity.Id);
+            }
+
+            try
+            {
+                await context.Invoke();
+                if (activity is not null && activity.IsAllDataRequested)
+                {
+                    activity.SetTag("status", "Ok");
+                }
+            }
+            catch (Exception e)
+            {
+                if (activity is not null && activity.IsAllDataRequested)
+                {
+                    // exception attributes from https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/semantic_conventions/exceptions.md
+                    activity.SetTag("exception.type", e.GetType().FullName);
+                    activity.SetTag("exception.message", e.Message);
+                    activity.SetTag("exception.stacktrace", e.StackTrace);
+                    activity.SetTag("exception.escaped", true);
+                    activity.SetTag("status", "Error");
+                }
+                throw;
+            }
+        }
+    }
+
+    internal class ActivityPropagationOutgoingGrainCallFilter : ActivityPropagationGrainCallFilter, IOutgoingGrainCallFilter
+    {
+        public Task Invoke(IOutgoingGrainCallContext context)
+        {
+            if (Activity.Current != null)
+            {
+                return ProcessCurrentActivity(context); // Copy existing activity to RequestContext
+            }
+            return ProcessNewActivity(context, ActivityNameOut, ActivityKind.Client, new ActivityContext()); // Create new activity
+        }
+
+        private static Task ProcessCurrentActivity(IOutgoingGrainCallContext context)
+        {
+            var currentActivity = Activity.Current;
+
+            if (currentActivity is not null &&
+                currentActivity.IdFormat == ActivityIdFormat.W3C)
+            {
+                RequestContext.Set(TraceParentHeaderName, currentActivity.Id);
+                if (currentActivity.TraceStateString is not null)
+                    RequestContext.Set(TraceStateHeaderName, currentActivity.TraceStateString);
+            }
```

Comment:
As I mentioned on the incoming filter, the outgoing filter should also use a propagator.